File: crates/fortress-js/src/policy.ts
/**
 * Policy engine implementation for Fortress JavaScript/TypeScript SDK
 */

import { PolicyConfig, PolicyRule, Condition, Role, Permission, Resource } from './types';

export class PolicyEngine {
  private config: PolicyConfig;
  private roles: Map<string, Role>;
  private rules: Map<string, PolicyRule>;

  constructor(config?: PolicyConfig) {
    this.config = config || {
      name: 'default',
      version: '1.0',
      rules: [],
      default_action: 'deny',
      evaluation_mode: 'all'
    };
    this.roles = new Map();
    this.rules = new Map();
  }

  async createRole(role: Role): Promise<string> {
    this.roles.set(role.id, role);
    return role.id;
  }

  async deleteRole(roleId: string): Promise<void> {
    if (!this.roles.has(roleId)) {
      throw new Error(`Role not found: ${roleId}`);
    }
    this.roles.delete(roleId);
  }

  async getRole(roleId: string): Promise<Role> {
    const role = this.roles.get(roleId);
    if (!role) {
      throw new Error(`Role not found: ${roleId}`);
    }
    return role;
  }

  async listRoles(): Promise<Role[]> {
    return Array.from(this.roles.values());
  }

  async updateRole(roleId: string, updates: Partial<Role>): Promise<void> {
    const role = this.roles.get(roleId);
    if (!role) {
      throw new Error(`Role not found: ${roleId}`);
    }
    
    const updatedRole = { ...role, ...updates };
    this.roles.set(roleId, updatedRole);
  }

  async createRule(rule: PolicyRule): Promise<string> {
    this.rules.set(rule.id, rule);
    return rule.id;
  }

  async deleteRule(ruleId: string): Promise<void> {
    if (!this.rules.has(ruleId)) {
      throw new Error(`Rule not found: ${ruleId}`);
    }
    this.rules.delete(ruleId);
  }

  async getRule(ruleId: string): Promise<PolicyRule> {
    const rule = this.rules.get(ruleId);
    if (!rule) {
      throw new Error(`Rule not found: ${ruleId}`);
    }
    return rule;
  }

  async listRules(): Promise<PolicyRule[]> {
    return Array.from(this.rules.values());
  }

  async evaluatePolicy(
    roleId: string,
    action: string,
    resource: Resource,
    context?: any
  ): Promise<{ allowed: boolean; reason?: string; matchedRules: string[] }> {
    const role = this.roles.get(roleId);
    if (!role) {
      return {
        allowed: false,
        reason: `Role not found: ${roleId}`,
        matchedRules: []
      };
    }

    const matchedRules: string[] = [];
    let allowCount = 0;
    let denyCount = 0;

    // Check role permissions first
    for (const permission of role.permissions) {
      if (this.matchesResource(permission.resource, resource) &&
          this.matchesAction(permission.actions, action)) {
        
        if (this.evaluateConditions(permission.conditions, context)) {
          matchedRules.push(`role:${role.id}:${permission.id}`);
          
          if (permission.effect === 'allow') {
            allowCount++;
          } else {
            denyCount++;
          }
        }
      }
    }

    // Check policy rules
    for (const rule of this.rules.values()) {
      if (rule.enabled && this.evaluateRuleConditions(rule.conditions, context)) {
        matchedRules.push(`rule:${rule.id}`);
        
        if (rule.action === 'allow') {
          allowCount++;
        } else {
          denyCount++;
        }
      }
    }

    // Evaluate based on evaluation mode
    let allowed = false;
    let reason: string | undefined;

    if (this.config.evaluation_mode === 'all') {
      allowed = allowCount > 0 && denyCount === 0;
      if (!allowed) {
        reason = denyCount > 0 ? 'Denied by explicit deny rule' : 'No matching allow rules found';
      }
    } else { // 'any' mode
      allowed = allowCount > 0;
      if (!allowed) {
        reason = 'No matching allow rules found';
      }
    }

    return { allowed, reason, matchedRules };
  }

  private matchesResource(permissionResource: Resource, targetResource: Resource): boolean {
    if (permissionResource.id && targetResource.id) {
      return permissionResource.id === targetResource.id;
    }
    
    if (permissionResource.type && targetResource.type) {
      return permissionResource.type === targetResource.type;
    }
    
    return false;
  }

  private matchesAction(permissionActions: string[], targetAction: string): boolean {
    return permissionActions.includes('*') || permissionActions.includes(targetAction);
  }

  private evaluateConditions(conditions: Condition[] | undefined, context: any): boolean {
    if (!conditions || conditions.length === 0) {
      return true;
    }

    return conditions.every(condition => this.evaluateCondition(condition, context));
  }

  private evaluateRuleConditions(conditions: Condition[] | undefined, context: any): boolean {
    if (!conditions || conditions.length === 0) {
      return true;
    }

    return conditions.every(condition => this.evaluateCondition(condition, context));
  }

  private evaluateCondition(condition: Condition, context: any): boolean {
    if (!context) {
      return true;
    }

    const fieldValue = this.getFieldValue(context, condition.field);
    const conditionValue = condition.value;

    switch (condition.operator) {
      case 'eq':
        return fieldValue === conditionValue;
      case 'ne':
        return fieldValue !== conditionValue;
      case 'gt':
        return Number(fieldValue) > Number(conditionValue);
      case 'gte':
        return Number(fieldValue) >= Number(conditionValue);
      case 'lt':
        return Number(fieldValue) < Number(conditionValue);
      case 'lte':
        return Number(fieldValue) <= Number(conditionValue);
      case 'in':
        return Array.isArray(conditionValue) && conditionValue.includes(fieldValue);
      case 'contains':
        return String(fieldValue).includes(String(conditionValue));
      default:
        return false;
    }
  }

  private getFieldValue(obj: any, field: string): any {
    return field.split('.').reduce((current, key) => current?.[key], obj);
  }

  async assignRoleToUser(userId: string, roleId: string): Promise<void> {
    // In a real implementation, this would store the user-role mapping
    // For now, we'll just validate the role exists
    const role = this.roles.get(roleId);
    if (!role) {
      throw new Error(`Role not found: ${roleId}`);
    }
  }

  async removeRoleFromUser(userId: string, roleId: string): Promise<void> {
    // In a real implementation, this would remove the user-role mapping
  }

  async getUserRoles(userId: string): Promise<string[]> {
    // In a real implementation, this would fetch user roles from storage
    return [];
  }

  async checkPermission(
    userId: string,
    resource: Resource,
    action: string,
    context?: any
  ): Promise<{ allowed: boolean; reason?: string }> {
    const userRoles = await this.getUserRoles(userId);
    
    for (const roleId of userRoles) {
      const result = await this.evaluatePolicy(roleId, action, resource, context);
      if (result.allowed) {
        return { allowed: true };
      }
    }

    return { allowed: false, reason: 'No permissions granted' };
  }
}

File: crates/fortress-go/policy.go
package fortress

import (
	"fmt"
	"sync"
	"time"
)

// PolicyEngine manages access policies
type PolicyEngine struct {
	config  *Config
	policies map[string]*PolicyConfig
	mu      sync.RWMutex
}

// NewPolicyEngine creates a new policy engine
func NewPolicyEngine(config *Config) *PolicyEngine {
	return &PolicyEngine{
		config:  config,
		policies: make(map[string]*PolicyConfig),
	}
}

// AddPolicy adds a policy
func (p *PolicyEngine) AddPolicy(policy *PolicyConfig) {
	p.mu.Lock()
	defer p.mu.Unlock()

	p.policies[policy.Name] = policy
}

// RemovePolicy removes a policy
func (p *PolicyEngine) RemovePolicy(name string) bool {
	p.mu.Lock()
	defer p.mu.Unlock()

	_, exists := p.policies[name]
	if exists {
		delete(p.policies, name)
		return true
	}
	return false
}

// GetPolicy retrieves a policy by name
func (p *PolicyEngine) GetPolicy(name string) *PolicyConfig {
	p.mu.RLock()
	defer p.mu.RUnlock()

	policy, exists := p.policies[name]
	if !exists {
		return nil
	}
	return policy
}

// ListPolicies returns all policy names
func (p *PolicyEngine) ListPolicies() []string {
	p.mu.RLock()
	defer p.mu.RUnlock()

	var names []string
	for name := range p.policies {
		names = append(names, name)
	}
	return names
}

// EvaluatePermission evaluates if a user has permission for an action on a resource
func (p *PolicyEngine) EvaluatePermission(userID string, resource *Resource, action string, context map[string]interface{}) (bool, error) {
	p.mu.RLock()
	policies := make([]*PolicyConfig, 0, len(p.policies))
	for _, policy := range p.policies {
		policies = append(policies, policy)
	}
	p.mu.RUnlock()

	for _, policy := range policies {
		result, err := p.evaluatePolicy(policy, userID, resource, action, context)
		if err != nil {
			return false, NewPolicyError("policy evaluation failed", err)
		}

		if policy.EvaluationMode == "any" && result {
			return true, nil
		}

		if policy.EvaluationMode == "all" && !result {
			return false, nil
		}
	}

	// Return default action if no policies matched
	return p.config.Policy.DefaultAction == "allow", nil
}

// evaluatePolicy evaluates a single policy
func (p *PolicyEngine) evaluatePolicy(policy *PolicyConfig, userID string, resource *Resource, action string, context map[string]interface{}) (bool, error) {
	// Sort rules by priority (higher priority first)
	sortedRules := make([]*PolicyRule, len(policy.Rules))
	copy(sortedRules, policy.Rules)

	// Simple sort by priority (higher first)
	for i := 0; i < len(sortedRules)-1; i++ {
		for j := i + 1; j < len(sortedRules); j++ {
			if sortedRules[i].Priority < sortedRules[j].Priority {
				sortedRules[i], sortedRules[j] = sortedRules[j], sortedRules[i]
			}
		}
	}

	for _, rule := range sortedRules {
		if !rule.Enabled {
			continue
		}

		matches, err := p.evaluateRule(rule, userID, resource, action, context)
		if err != nil {
			return false, err
		}

		if matches {
			return rule.Action == "allow", nil
		}
	}

	return policy.DefaultAction == "allow", nil
}

// evaluateRule evaluates a single rule
func (p *PolicyEngine) evaluateRule(rule *PolicyRule, userID string, resource *Resource, action string, context map[string]interface{}) (bool, error) {
	// Check if action matches
	permissionMatches := false
	for _, permission := range rule.Permissions {
		if permission.Resource.Type == resource.Type && 
		   (permission.Resource.ID == "" || permission.Resource.ID == resource.ID) {
			for _, allowedAction := range permission.Actions {
				if allowedAction == action {
					permissionMatches = true
					break
				}
			}
		}
	}

	if !permissionMatches {
		return false, nil
	}

	// Evaluate conditions
	for _, condition := range rule.Conditions {
		matches, err := p.evaluateCondition(condition, userID, resource, action, context)
		if err != nil {
			return false, err
		}

		if !matches {
			return false, nil
		}
	}

	return true, nil
}

// evaluateCondition evaluates a single condition
func (p *PolicyEngine) evaluateCondition(condition Condition, userID string, resource *Resource, action string, context map[string]interface{}) (bool, error) {
	fieldValue := p.getFieldValue(condition.Field, userID, resource, action, context)

	switch condition.Operator {
	case "eq":
		return p.compareValues(fieldValue, condition.Value, "==")
	case "ne":
		return !p.compareValues(fieldValue, condition.Value, "==")
	case "gt":
		return p.compareValues(fieldValue, condition.Value, ">")
	case "gte":
		return p.compareValues(fieldValue, condition.Value, ">=")
	case "lt":
		return p.compareValues(fieldValue, condition.Value, "<")
	case "lte":
		return p.compareValues(fieldValue, condition.Value, "<=")
	case "in":
		return p.valueInArray(fieldValue, condition.Value)
	case "contains":
		return p.stringContains(fieldValue, condition.Value)
	default:
		return false, NewPolicyError(fmt.Sprintf("unknown operator: %s", condition.Operator), nil)
	}
}

// getFieldValue gets the value of a field for evaluation
func (p *PolicyEngine) getFieldValue(field string, userID string, resource *Resource, action string, context map[string]interface{}) interface{} {
	switch field {
	case "user.id":
		return userID
	case "resource.type":
		return resource.Type
	case "resource.id":
		return resource.ID
	case "action":
		return action
	default:
		// Try to get from context
		if context != nil {
			if value, exists := context[field]; exists {
				return value
			}
		}
		return nil
	}
}

// compareValues compares two values based on an operator
func (p *PolicyEngine) compareValues(fieldValue, conditionValue interface{}, operator string) bool {
	switch operator {
	case "==":
		return fieldValue == conditionValue
	case ">":
		return p.compareNumeric(fieldValue, conditionValue, func(a, b float64) bool { return a > b })
	case ">=":
		return p.compareNumeric(fieldValue, conditionValue, func(a, b float64) bool { return a >= b })
	case "<":
		return p.compareNumeric(fieldValue, conditionValue, func(a, b float64) bool { return a < b })
	case "<=":
		return p.compareNumeric(fieldValue, conditionValue, func(a, b float64) bool { return a <= b })
	default:
		return false
	}
}

// compareNumeric compares two numeric values
func (p *PolicyEngine) compareNumeric(a, b interface{}, compareFunc func(float64, float64) bool) bool {
	aFloat, aOk := p.toFloat64(a)
	bFloat, bOk := p.toFloat64(b)

	if !aOk || !bOk {
		return false
	}

	return compareFunc(aFloat, bFloat)
}

// toFloat64 converts a value to float64
func (p *PolicyEngine) toFloat64(value interface{}) (float64, bool) {
	switch v := value.(type) {
	case int:
		return float64(v), true
	case int64:
		return float64(v), true
	case float32:
		return float64(v), true
	case float64:
		return v, true
	default:
		return 0, false
	}
}

// valueInArray checks if a value exists in an array
func (p *PolicyEngine) valueInArray(value, array interface{}) bool {
	arr, ok := array.([]interface{})
	if !ok {
		return false
	}

	for _, item := range arr {
		if p.compareValues(value, item, "==") {
			return true
		}
	}
	return false
}

// stringContains checks if a string contains another string
func (p *PolicyEngine) stringContains(value, substr interface{}) bool {
	valueStr, ok := value.(string)
	substrStr, ok := substr.(string)

	if !ok || !ok {
		return false
	}

	for i := 0; i <= len(valueStr)-len(substrStr); i++ {
		if valueStr[i:i+len(substrStr)] == substrStr {
			return true
		}
	}
	return false
}

// CreatePermission creates a new permission
func (p *PolicyEngine) CreatePermission(resource *Resource, actions []string) *Permission {
	return &Permission{
		ID:       GenerateRandomString(8),
		Resource: *resource,
		Actions:  actions,
	}
}

// CreateRule creates a new policy rule
func (p *PolicyEngine) CreateRule(name string, conditions []Condition, action string, priority int) *PolicyRule {
	return &PolicyRule{
		ID:         GenerateRandomString(8),
		Name:       name,
		Conditions: conditions,
		Action:     action,
		Priority:   priority,
		Enabled:    true,
	}
}
